Add basic support for RabbitMQ High Availability connections

mihaitodor opened the pull request:
This is required by https://github.com/upfluence/sensu-client-go/issues/13

Limitations / TODOs for future PRs:
- Prefetch values are currently ignored
- No SSL support
- Documentation enhances

Note: The test for `RabbitMQTransport.Connect` is very basic. Doing more tests requires either adding a wrapper for the amqp functions or possibly using something like [this package](https://github.com/bouk/monkey). Please let me know what are your thoughts on this.

Note 2: I'm reusing a function from rabbitmq_config_test.go in rabbitmq_test.go, which is definitely ugly. I also copied [this function](https://github.com/upfluence/sensu-client-go/blob/master/sensu/config_test.go#L19) from sensu-client-go. I'd rather have these functions in some common place before merging the code. Could you please suggest where I should place them? Perhaps some extra folder in this library?

mihaitodor committed:
Add basic support for RabbitMQ High Availability

mihaitodor committed:
Fix some issues reported by the linter

mihaitodor committed:
Add RabbitMQ TransportConfig tests

mihaitodor committed:
Fix heartbeat condition

mihaitodor committed:
Add RabbitMQTransport tests

mihaitodor committed:
Fix err variable shadowing

AlexisMontagne commented on sensu/transport/rabbitmq/rabbitmq.go:
```diff
@@ -2,30 +2,84 @@ package rabbitmq
 
-	"github.com/upfluence/sensu-go/Godeps/_workspace/src/github.com/streadway/amqp"
+	"github.com/upfluence/sensu-client-go/Godeps/_workspace/src/github.com/upfluence/goutils/log"
+
```
you can get rid of this new line

AlexisMontagne commented on sensu/transport/rabbitmq/rabbitmq.go:
```diff
@@ -2,30 +2,84 @@ package rabbitmq
 }
 
 func (t *RabbitMQTransport) Connect() error {
+	var uri string
```
when you have multiple assignments you can use a multiple line `var` statement:

```
var (
   uri string
   err string

   randGenerator = rand.New(rand.NewSource(time.Now().UnixNano()))
)
```

mihaitodor replied:
Ah, I missed that... 👍

AlexisMontagne commented on sensu/transport/rabbitmq/rabbitmq_config.go:
```diff
@@ -0,0 +1,80 @@
+}
+
+func (c *TransportConfig) GetURI() string {
+	if c.uri == "" {
```
Does this "caching" layer is really useful?

mihaitodor replied:
Bad habits die hard :)

AlexisMontagne commented on sensu/transport/rabbitmq/rabbitmq.go:
```diff
@@ -2,30 +2,84 @@ package rabbitmq
+		heartbeatString := config.Heartbeat.String()
+		if heartbeatString != "" {
+			var heartbeat time.Duration
+			heartbeat, err = time.ParseDuration(heartbeatString)
```
use the type inference  assignation: `heartbeat, err := time.ParseDuration(heartbeatString)`

AlexisMontagne commented on sensu/transport/rabbitmq/rabbitmq.go:
```diff
@@ -2,30 +2,84 @@ package rabbitmq
+				continue
+			}
+
+			t.Connection, err = amqp.DialConfig(uri, amqp.Config{
```
Go to the new line after `(`

```
t.Connection, err = amqp.DialConfig(
   uri, 
   amqp.Config{Heartbeat: heartbeat},
)
```

AlexisMontagne commented on sensu/transport/rabbitmq/rabbitmq_config.go:
```diff
@@ -0,0 +1,80 @@
+func NewTransportConfig(uri string) (*TransportConfig, error) {
+	uriComponents, err := url.Parse(uri)
+	if err != nil {
+		log.Errorf("Failed to parse the URI: %s", err)
```
is it really useful to log the errors before return them? I do think it would be easier to only log the error returned when you call `NewTransportConfig`

mihaitodor replied:
Agreed, logging is pointless there

AlexisMontagne commented on sensu/transport/rabbitmq/rabbitmq.go:
```diff
@@ -2,30 +2,84 @@ package rabbitmq
+	config, err := NewTransportConfig(uri)
+
+	if err != nil {
+		log.Errorf("Received invalid URI %s", err)
```
cf. the comment in `NewTransportConfig`

mihaitodor committed:
Fix bug: time.ParseDuration fails to parse duration

- time.ParseDuration requires an 's' at the end of the string for
the heartbeat parameter

mihaitodor committed:
Remove useless space from imports

mihaitodor committed:
Remove useless URI caching

mihaitodor committed:
Remove superfluous logging and update unit tests

mihaitodor committed:
Fix formatting issues

AlexisMontagne commented on sensu/transport/rabbitmq/rabbitmq.go:
```diff
@@ -2,30 +2,86 @@ package rabbitmq
+	"time"
 
-	"github.com/upfluence/sensu-go/Godeps/_workspace/src/github.com/streadway/amqp"
+	"github.com/upfluence/sensu-client-go/Godeps/_workspace/src/github.com/streadway/amqp"
```
the path reference the dependencies vendored in `github.com/upfluence/sensu-client-go`, it should be referencing the one in `github.com/upfluence/sensu-go`

mihaitodor replied:
Ouch... Not sure how I messed that one up. I'll blame it on goimports for now and pay more attention next time.

mihaitodor committed:
Fix messed-up imports

AlexisMontagne commented:
It looks good to me!

About your notes:

Note 1: What you can do is using interfaces instead of actual structs ( `amqp.Connection`  / `amqp.Channel` ) and abstract the `amqp.DialConfig` method as a member of struct: 

```golang
type  amqpChannel interface {
   ....
}

type  amqpConnection interface {
    Channel() (amqpChannel, error)
}

type RabbitMQTransport struct {
     connection amqpConnection 
     channel amqpChannel
     dialer func(string, amqp.Config) (amqpConnection, error)
}
```

Note 2:

Feel free to create a package: `github.com/upfluence/sensu-go/testing` and put the helpers in it.
